sympy/assumptions/handlers/order.py

```diff
@@ -66,32 +66,88 @@ def Add(expr, assumptions):
         if r is not True:
             return r
 
-        nonpos = 0
-        for arg in expr.args:
-            if ask(Q.negative(arg), assumptions) is not True:
-                if ask(Q.positive(arg), assumptions) is False:
-                    nonpos += 1
-                else:
-                    break
-        else:
-            if nonpos < len(expr.args):
-                return True
+
+        args = [arg for arg in expr.args if not ask(Q.zero(arg), assumptions)]
+        if not args:
+            return False
+
+        inf = neg_inf = pos_inf = neg = npos = nneg = False
+
+        for arg in args:
+            if pos_inf and neg_inf:
+                return None
+
+            _inf = ask(Q.infinite(arg), assumptions)
+            if _inf:
+                inf = True
+
+            n = ask(Q.negative(arg), assumptions)
+            if n:
+                if _inf is not False:
+                    neg_inf = True
+                neg = True
+                continue
+
+            np = ask(Q.nonpositive(arg), assumptions)
+            if np:
+                if _inf is True:
+                    neg_inf = True
+                npos = True
+                continue
+
+            nn = ask(Q.nonnegative(arg), assumptions)
+            if nn:
+                if _inf or (_inf is None and nn is False):
+                    pos_inf = True
+                nneg = True
+                continue
+            else:
+                return None
```

Comment:
Why is it necessary to ask both `negative` and `nonnegative`? Isn't `n = fuzzy_not(nn)`?

Reply:
Are these guaranteed to be real and finite? If not then it is possible for both negative and nonnegative to be False. In the old assumptions:
```julia
In [1]: sqrt(-1).is_negative                                                                                                      
Out[1]: False

In [2]: sqrt(-1).is_nonnegative                                                                                                   
Out[2]: False

In [3]: oo.is_negative                                                                                                            
Out[3]: False

In [4]: oo.is_nonnegative                                                                                                         
Out[4]: False
```

Reply:
This function starts with a check for real numbers, so I think that's fine. I assumed that sympy distinguished positive and negative infinity, but maybe I was wrong.

Reply:
Yup, even I had the same doubt but it is okay.

Reply:
Positive and negative infinity are distinguished. The point is that `is_negative` means "finite and negative". We have extended versions of these that can include the infinities:
```julia
In [16]: oo.is_real                                                                                                               
Out[16]: False

In [17]: oo.is_extended_real                                                                                                      
Out[17]: True

In [18]: oo.is_extended_positive                                                                                                  
Out[18]: True

In [19]: oo.is_extended_negative                                                                                                  
Out[19]: False
```
The surprising result is this:
```julia
In [20]: oo.is_positive                                                                                                           
Out[20]: False
```
That's because positive implies real which implies finite. The full list of implications in the old assumption system is here:
https://github.com/sympy/sympy/blob/94fb720696f5f5d12bad8bc813699fd696afd2fb/sympy/core/assumptions.py#L203-L260

For the new assumptions it is here:
https://github.com/sympy/sympy/blob/94fb720696f5f5d12bad8bc813699fd696afd2fb/sympy/assumptions/ask.py#L1467-L1520